### PR TITLE
add static replica dividing

### PR DESCRIPTION
**Pull request by Garrybest**
#### What type of PR is this?

kind/feature
kind/api
kind/change

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: pkg/scheduler/algorithm/generic.go [hunk 1]
```diff
@@ -64,7 +65,7 @@ type genericScheduler struct {
 // Schedule tries to schedule the given subscription to multiple clusters.
 // If it succeeds, it will return the namespaced names of ManagedClusters.
 // If it fails, it will return a FitError error with reasons.
-func (g *genericScheduler) Schedule(ctx context.Context, fwk framework.Framework, sub *appsapi.Subscription) (result ScheduleResult, err error) {
+func (g *genericScheduler) Schedule(ctx context.Context, fwk framework.Framework, sub *appsapi.Subscription, finv *appsapi.FeedInventory) (result ScheduleResult, err error) {
```
Comment (dixudx) [inline pkg/scheduler/algorithm/generic.go:68]: `finv` is only used in `selectClusters`. I don't think we need to pass it down from framework interface.

File: pkg/scheduler/algorithm/interface.go [hunk 2]
```diff
@@ -26,14 +26,14 @@ import (
 // ScheduleAlgorithm is an interface implemented by things that know how to schedule resources to target
 // managed clusters.
 type ScheduleAlgorithm interface {
-	Schedule(context.Context, framework.Framework, *appsapi.Subscription) (scheduleResult ScheduleResult, err error)
+	Schedule(context.Context, framework.Framework, *appsapi.Subscription, *appsapi.FeedInventory) (scheduleResult ScheduleResult, err error)
```
Comment (dixudx) [inline pkg/scheduler/algorithm/interface.go:29]: Same as above.

No need to append new field `FeedInventory`.

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -231,34 +237,44 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
...
+	var finv *appsapi.FeedInventory
+	if sub.Spec.SchedulingStrategy == appsapi.DividingSchedulingStrategyType {
+		finv, err = sched.inventoryLister.FeedInventories(ns).Get(name)
+		// TODO(Garrybest): Ignore some circumstances when subscription does not need to be scheduled so far.
+		// E.g., when feed in subscription has been modified but finv has not been updated.
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				utilruntime.HandleError(err)
+			}
+			return
```
Comment (dixudx) [inline pkg/scheduler/scheduler.go:249]: What if the corresponding `FeedInventory` is not created?

We should enqueue and re-schedule later.

Comment (Garrybest) [reply pkg/scheduler/scheduler.go:249]: I think the `FeedInventory` event handler will add the subcription. Please check the `addAllEventHandlers` function.

File: pkg/scheduler/algorithm/generic.go [hunk 4]
```diff
@@ -103,16 +104,23 @@ func (g *genericScheduler) Schedule(ctx context.Context, fwk framework.Framework
...
-	var selected []string
+	var selected framework.TargetClusters
 	for _, clusterScore := range clusterScoreList {
 		// TODO: sampling with scores
-		selected = append(selected, clusterScore.NamespacedName)
+		selected.BindingClusters = append(selected.BindingClusters, clusterScore.NamespacedName)
 	}
+
+	if sub.Spec.SchedulingStrategy == appsapi.DividingSchedulingStrategyType {
+		if err := g.divideReplicas(&selected, sub, finv); err != nil {
```
Comment (dixudx) [inline pkg/scheduler/algorithm/generic.go:119]: `selected` will keep redundant clusters that are not bound for dividing scheduling.

Leave a knot here.

File: pkg/scheduler/algorithm/generic_test.go [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+	return &appsapi.Feed{
+		Kind:       "Deployment",
+		APIVersion: "apps/v1",
+		Namespace:  "default",
+		Name:       name,
+	}
+}
+
+func Test_genericScheduler_divideReplicas(t *testing.T) {
+	weight0, weight1, weight2 := int32(0), int32(1), int32(2)
```
Comment (dixudx) [inline pkg/scheduler/algorithm/generic_test.go:42]:
```
utilpointer "k8s.io/utils/pointer"
```
could use `utilpointer.Int32(1)` instead.

**Review by dixudx — APPROVED** [review]

**Pull request merged**